### ISSUE TITLE
Status on Frida on musl-based linux platforms?

### ISSUE THREAD

**Issue by luckycat889** [frida/frida#2787]
Hello,
I read some release logs about the addition of musl support not long ago in version 16.0.10.
Quick question; are there any outstanding issues regarding the musl support or is it expected to work as well as on glibc systems? I see some HAVE_MUSL ifdefs, which I am not sure I fully understand....

Background; I want to use frida on a system with a really old musl. I see in the GUM code that it appears that the logic will check for presence of 'dl_iterate_phdr' and will fall back reading the /proc file system, when a listing of loaded shared object is requested. In my case, I see that my musl libc.so does not include dl_iterate_phdr (yet), and I wonder if it's expected to fallback, with full functionality, to the /proc method, or am I hosed?

Thanks in advance!

**Comment by oleavr** [comment]
Hey! Yes, there aren't any known issues -- only drawback compared to glibc is that we have no way of unloading frida-agent, so it sticks around indefinitely. The fallback in Gum is indeed fine, we don't require the libc to support that API. Cheers!

**Comment by luckycat889** [comment]
Appreciate the quick answer; good to know regarding musl.

So, in the old musl system I mentioned, I have built the latest from source and I try the regular:
```
-bash-5.2# frida-trace -i "recv*" -i "read*" <proc>
Failed to attach: error relocating /proc/self/fd/3: dl_iterate_phdr: symbol not found
-bash-5.2# 
```
I see where the said fallback is happening, and in gumprocess-linux.c, I believe this is the only place where dl_iterate_phdr is being used. Does anyone know what's going on? I tried to make a small mod to force usage of /proc, but it still reports the same error.
Thanks!

**Comment by oleavr** [comment]
Ooh, the issue is probably in libunwind. You can rebuild it like this:

```sh
$ make -f Makefile.sdk.mk libunwind FRIDA_HOST=…
```

And then symlink it into your build/sdk-… like this:

```sh
$ make -f Makefile.sdk.mk symlinks-libunwind FRIDA_HOST=…
```

After that make a dummy-change like adding a blank line to `frida-core/lib/agent/agent-glue.c`, to force a relink of `frida-agent.so`, and hopefully that should take care of it. If not, you might have to build more pieces of the SDK. Omit `libunwind` from the first example to build all of it.

**Comment by luckycat889** [comment]
I hate to pester regarding this....
Previously I had everything built from scratch, including SDK. But I took your point and rebuilt everything using a musl toolchain, which is as close to what the target binaries were built with (I made sure my toolchain uses musl 1.1.6, which I know the target binaries are loading).  For clarity, I didn't build the target binaries, and I am just doing my best figuring out how they were built.
But still I get the missing dl_iterate_phdr, as described earlier.
I understand what you are saying with the fallback, and I will tinker regarding this and if I get something working and it's usable, I send in an MR.

However, I just want to make sure about one thing; I basically copy everything in build/frida-linux-x86-musl onto the target. I gather that libfrida-agent.so is somehow embedded into the executables/environment under build/frida-linux-x86-musl (since libfrida-agent.so is only residing in the build/tmp-linux-x86-musl/frida-core/lib/agent folder)? Or did I make an incorrect assumption regarding what I need to copy to target?

For good measure, I built some other programs for the target, for instance 'htop', and I test to trace with frida. I launch 'htop' in a separate terminal and I run:
```
-bash-5.2# frida-trace --decorate -i "recv*" -i "send*" htop
Failed to attach: process with pid 227 either refused to load frida-agent, or terminated during injection

```
The htop process dies and based on the output I asked the question about the agent. If you have any ideas, that would be appreciated. A difference between htop and the initial target binary is that htop currently is using the exact libc.so from my toolchain (I added an RPATH to the binary, later though I will try to make use of the exact target resident libc.so)

So to be clear, I tried frida-trace on

- target resident binaries, and it's looking for the dl_iterate_phdr
- my own binaries, and I get the crash + complaint about agent
